From 9fb98a1ae8bef46c0a346cd3afe333eb56b7f799 Mon Sep 17 00:00:00 2001
From: Saleemuddin Shaik <quic_saleemud@quicinc.com>
Date: Thu, 17 Aug 2023 16:09:50 +0530
Subject: [PATCH] ath12k: add support for low_latency_rate_cnt

By accessing HTT_STATS_TX_PDEV_SAWF_RATE_STATS_TAG,
can able to view the values of low_latency_rate_cnt,

When the service class contains delay bound rate parameters which
indicate low latency and we enable latency-based RA params then
the low_latency_rate_count will be incremented.

commands:
root@OpenWrt:/#
echo 9 > /sys/kernel/debug/ath12k/qcn9274 hw2.0_0004:01:00.0/mac0/htt_stats_type

root@OpenWrt:/#
cat /sys/kernel/debug/ath12k/qcn9274 hw2.0_0004:01:00.0/mac0/htt_stats

output:
HTT_TX_PDEV_RATE_STATS_TLV:
mac_id = 0
tx_ldpc = 95675
ac_mu_mimo_tx_ldpc = 0
ax_mu_mimo_tx_ldpc = 0
ofdma_tx_ldpc = 0
rts_cnt = 129847
rts_success = 90133
ack_rssi = 96
Legacy CCK Rates: 1 Mbps: 0, 2 Mbps: 0, 5.5 Mbps: 0, 11 Mbps: 0
Legacy OFDM Rates: 6 Mbps: 287158, 9 Mbps: 0, 12 Mbps: 0, 18 Mbps: 0
                   24 Mbps: 0, 36 Mbps: 0, 48 Mbps: 0, 54 Mbps: 0
tx_mcs =  0:5212, 1:668, 2:751, 3:1016, 4:2729, 5:7208, 6:8332, 7:14312, 8:28652, 9:35729, 10:0, 11:0,
tx_mcs_ext =  0:0, 1:0,
ac_mu_mimo_tx_mcs =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
ac_mu_mimo_tx_mcs_ext =  0:0, 1:0,
ax_mu_mimo_tx_mcs =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
ax_mu_mimo_tx_mcs_ext =  0:0, 1:0,
ofdma_tx_mcs =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
ofdma_tx_mcs_ext =  0:0, 1:0,
tx_nss =  0:8394, 1:96215, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0,
ac_mu_mimo_tx_nss =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0,
ax_mu_mimo_tx_nss =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0,
ofdma_tx_nss =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0,
tx_bw =  0:16781, 1:4202, 2:83626, 3:0, 4:0,
ac_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0,
ax_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0,
ofdma_tx_bw =  0:0, 1:0, 2:0, 3:0,
tx_stbc =  0:0, 1:0, 2:0, 3:0, 4:0, 5:19, 6:82, 7:69, 8:46, 9:47, 10:0, 11:0,
tx_pream =  0:287398, 1:0, 2:13271, 3:145690, 4:0, 5:0, 6:0,
HE LTF: 1x: 0, 2x: 0, 4x: 0
tx_gi[0] =
 -2:0,-1:0, 0:4965, 1:376, 2:265, 3:265, 4:545, 5:1228, 6:1461, 7:702, 8:1829, 9:1862, 10:0, 11:0,
 0:0, 1:0,
tx_gi[1] =
 -2:0,-1:0, 0:247, 1:292, 2:486, 3:751, 4:2184, 5:5980, 6:6871, 7:13610, 8:26823, 9:33867, 10:0, 11:0,
 0:0, 1:0,
tx_gi[2] =
 -2:0,-1:0, 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
tx_gi[3] =
 -2:0,-1:0, 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ac_mu_mimo_tx_gi[0] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ac_mu_mimo_tx_gi[1] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ac_mu_mimo_tx_gi[2] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ac_mu_mimo_tx_gi[3] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ax_mu_mimo_tx_gi[0] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ax_mu_mimo_tx_gi[1] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ax_mu_mimo_tx_gi[2] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ax_mu_mimo_tx_gi[3] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ofdma_tx_gi[0] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ofdma_tx_gi[1] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ofdma_tx_gi[2] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ofdma_tx_gi[3] =
 0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0,
 0:0, 1:0,
ofdma_tx_ru_size =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0,
tx_dcm =  0:0, 1:0, 2:0, 3:0, 4:0,
tx_su_punctured_mode =  0:0, 1:0, 2:0, 3:0, 4:0,
half_tx_bw =  0:0, 1:0, 2:0, 3:0,
quarter_tx_bw =  0:0, 1:0, 2:0, 3:0,
half_ac_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0,
quarter_ac_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0,
half_ax_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0,
quarter_ax_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0,
half_ofdma_tx_bw 0:0, 1:0, 2:0, 3:0,
quarter_ofdma_tx_bw =  0:0, 1:0, 2:0, 3:0,
ofdma_he_sig_b_mcs =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0,
11ax_trigger_type =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0,
11be_trigger_type =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0,
ax_su_embedded_trigger_data_ppdu_cnt = 0
ax_su_embedded_trigger_data_ppdu_err_cnt = 0
HTT_TX_PDEV_BE_RATE_STATS_TLV:
be_mu_mimo_tx_ldpc = 0 -2:0,-1:0,be_mu_mimo_tx_mcs =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0,
be_mu_mimo_tx_nss =  0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0,
be_mu_mimo_tx_bw =  0:0, 1:0, 2:0, 3:0, 4:0,
be_mu_mimo_tx_gi[0] =   0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0,
be_mu_mimo_tx_gi[1] =   0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0,
be_mu_mimo_tx_gi[2] =   0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0,
be_mu_mimo_tx_gi[3] =   0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0,

low_latency_rate_cnt =  0
su_burst_rate_drop_cnt = 0
su_burst_rate_drop_fail_cnt = 0
PER_HISTOGRAM_STATS
Per_histogram_cnt:  0:85407, 1:1, 2:2, 3:47, 4:6, 5:7, 6:3, 7:16, 8:3, 9:3, 10:1, 11:8, 12:4, 13:4, 14:8, 15:6, 16:6, 17:4, 18:4, 19:2, 20:7, 21:2, 22:2, 23:2, 24:1,
  25:29, 26:3, 27:2, 28:4, 29:1, 30:4, 31:1, 32:2, 33:49, 34:0, 35:0, 36:2, 37:2, 38:1, 39:0, 40:3, 41:0, 42:1, 43:0, 44:1, 45:1, 46:3, 47:0, 48:1, 49:0,
  50:236, 51:1, 52:0, 53:0, 54:2, 55:1, 56:0, 57:0, 58:0, 59:1, 60:3, 61:2, 62:1, 63:0, 64:0, 65:2, 66:32, 67:0, 68:2, 69:2, 70:0, 71:0, 72:0, 73:1, 74:0,
  75:20, 76:0, 77:0, 78:0, 79:0, 80:0, 81:0, 82:0, 83:1, 84:0, 85:0, 86:0, 87:0, 88:1, 89:0, 90:1, 91:0, 92:0, 93:0, 94:1, 95:1, 96:2, 97:1, 98:0, 99:0, 100:18629,
HTT_TX_PDEV_PPDU_DUR_STATS_TLV:
pdev_up_time_us_low = 1750458097
pdev_up_time_us_high = 4
tx_success_time_us_low = 82403051
tx_success_time_us_high = 0
tx_fail_time_us_low = 39273504
tx_fail_time_us_high = 0
tx_ppdu_dur_hist_us_0 =  0-250 : 162412, 250-500 : 3165, 500-750 : 288, 750-1000 : 3756, 1000-1250 : 145, 1250-1500 : 209, 1500-1750 : 241, 1750-2000 : 2516,
tx_ppdu_dur_hist_us_1 =  2000-2250 : 204, 2250-2500 : 0, 2500-2750 : 0, 2750-3000 : 0, 3000-3250 : 0, 3250-3500 : 0, 3500-3750 : 0, 3750-4000 : 0,

Signed-off-by: Saleemuddin Shaik <quic_saleemud@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c | 9 +++++++++
 drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h | 3 +++
 2 files changed, 12 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
index b76865d..10c19c8 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
@@ -8088,6 +8088,15 @@ static inline void htt_print_histogram_stats_tlv(const void *tag_buf,
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
 	char data[HTT_HISTOGRAM_STATS_LEN] = {0};
 
+	len += scnprintf(buf + len, buf_len - len, "low_latency_rate_cnt =  %u\n",
+			 htt_stats_buf->low_latency_rate_cnt);
+
+	len += scnprintf(buf + len, buf_len - len, "su_burst_rate_drop_cnt = %u\n",
+			 htt_stats_buf->su_burst_rate_drop_cnt);
+
+	len += scnprintf(buf + len, buf_len - len, "su_burst_rate_drop_fail_cnt = %u\n",
+			 htt_stats_buf->su_burst_rate_drop_fail_cnt);
+
 	len += scnprintf(buf + len, buf_len - len, "PER_HISTOGRAM_STATS\n");
 
 	index = 0;
diff --git a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h
index 5f8f216..62bfacf 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h
+++ b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h
@@ -2556,6 +2556,9 @@ struct htt_tx_histogram_stats_tlv {
 	u32 rate_retry_mcs_drop_cnt;
 	u32 mcs_drop_rate[HTT_TX_PDEV_STATS_NUM_MCS_DROP_COUNTERS];
 	u32 per_histogram_cnt[HTT_TX_PDEV_STATS_NUM_PER_COUNTERS];
+	u32 low_latency_rate_cnt;
+	u32 su_burst_rate_drop_cnt;
+	u32 su_burst_rate_drop_fail_cnt;
 };
 
 struct htt_rx_pdev_ul_mumimo_trig_be_stats_tlv {
-- 
2.17.1

